### PR TITLE
metrics: fix bug triggered by memory reordering casualties

### PR DESCRIPTION
**Pull request by dianpopa**
With the passing of the time in Firecracker started
to shape out 2 types of metrics:
* the types of metrics needed for counting purposes -> they
are represented by 2 values, one of them is the active counter while
the other one is the value that was last flushed. With each flush, we would
 compute the delta between them
* the types of metrics needed for storing persistent values of a specific
measurement, for example the ones used for obtaining the process start up
time. This type of metric does not need to keep around a second value
since the value being flushed is always the one the system updates.

This approach also rids us of any memory reordering that may happen on weak
memory models.

Signed-off-by: Diana Popa <dpopa@amazon.com>

## Reason for This PR

The first are of code that needs to check out first thing when starting the review is `src/logger/metrics.rs`.

## Description of Changes

`[Author TODO: add description of changes.]`

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/logger/src/metrics.rs [hunk 1]
```diff
@@ -223,16 +237,19 @@ impl Metric for SharedMetric {
...
+impl StoreMetric for SharedStoreMetric {
+    fn fetch(&self) -> usize {
+        self.0.load(Ordering::Relaxed)
+    }
 
-    // It is necessary to also reset the old value in order to flush later the correct one,
-    // which is `value` (see the `Serialize` implementation for SharedMetric a few lines below).
     fn store(&self, value: usize) {
         self.0.store(value, Ordering::Relaxed);
-        self.1.store(0, Ordering::Relaxed)
```
Comment (acatangiu) [inline src/logger/src/metrics.rs:231]: Indeed this was racy between thread doing `store` and thread doing `serialize`, good catch!

**Review by acatangiu — COMMENTED** [review]
LGTM 👍

File: src/logger/src/metrics.rs [hunk 2]
```diff
@@ -185,7 +190,7 @@ impl fmt::Display for MetricsError {
 /// or an arbitrary amount of units.
 // This trait helps with writing less code. It has to be in scope (via an use directive) in order
```
Comment (alindima) [inline src/logger/src/metrics.rs:191]: I think we should also update this comment. Currently, it states the general purpose of traits, which is redundant. It should be similar to the comment associated to pub trait `StoreMetric`, stating the use case

File: src/logger/src/metrics.rs (hunk 1 above)
Comment (alindima) [reply src/logger/src/metrics.rs:231]: hmm, was this racy? I think the atomic type prevented that, right?

File: src/logger/src/metrics.rs [hunk 3]
```diff
@@ -774,16 +796,18 @@ mod tests {
...
+            metric.count(),
             M2_INITIAL_COUNT + NUM_THREADS_TO_SPAWN * NUM_INCREMENTS_PER_THREAD
         );
+    }
 
-        // Trying to store multiple values in the metric will result in keeping only the last
-        // value there.
-        m2.store(1);
-        m2.store(2);
-        assert_eq!(m2.0.load(Ordering::Relaxed), 2);
```
Comment (alindima) [inline src/logger/src/metrics.rs:785]: why did you remove this small test?

**Review by alindima — COMMENTED** [review]
Looks good. Only a couple of nits/questions

File: src/logger/src/metrics.rs (hunk 1 above)
Comment (alindima) [reply src/logger/src/metrics.rs:231]: makes total sense, thanks

File: src/logger/src/metrics.rs (hunk 3 above)
Comment (alindima) [reply src/logger/src/metrics.rs:785]: given the new implementation, I now see that this test doesn't make sense any more

File: src/logger/src/metrics.rs [hunk 4]
```diff
@@ -181,11 +186,9 @@ impl fmt::Display for MetricsError {
     }
 }
 
-/// Used for defining new types of metrics that can be either incremented with an unit
-/// or an arbitrary amount of units.
-// This trait helps with writing less code. It has to be in scope (via an use directive) in order
-// for its methods to be available to call on structs that implement it.
-pub trait Metric {
+/// Used for defining new types of metrics that as a counter (i.e they are continously updated by
```
Comment (alindima) [inline src/logger/src/metrics.rs:189]: nit:

```suggestion
/// Used for defining new types of metrics that act as a counter (i.e they are continously updated by
```

Comment (dianpopa) [reply src/logger/src/metrics.rs]: Done!

**Review by alindima — APPROVED** [review]

File: tests/integration_tests/build/test_binary_size.py [hunk 5]
```diff
@@ -13,9 +13,9 @@
 
 SIZES_DICT = {
     "x86_64": {
-        "FC_BINARY_SIZE_TARGET": 1955168,
+        "FC_BINARY_SIZE_TARGET": 2059824,
         "JAILER_BINARY_SIZE_TARGET": 1439512,
-        "FC_BINARY_SIZE_LIMIT": 2052927,
+        "FC_BINARY_SIZE_LIMIT": 2100000,
```
Comment (raduiliescu) [inline tests/integration_tests/build/test_binary_size.py:18]: Do we understand this increase? Can you detail it? just looking on the code a 2% increase doesn't make sense to me

Comment (alindima) [reply tests/integration_tests/build/test_binary_size.py:18]: Given the delta of 5%, the increase is not based only on this PR. The size has likely increased over time, and this PR was unlucky enough to go overboard the 5%

Comment (alindima) [reply tests/integration_tests/build/test_binary_size.py:18]: I saw this exact same issue on this PR #2223 , for ARM

Comment (raduiliescu) [reply tests/integration_tests/build/test_binary_size.py:18]: Ok.

**Review by raduiliescu — APPROVED** [review]
